File: MascotasUN/src/main/java/gestorAplicacion/gestionPersonas/CuentaBancaria.java
	/* Esta clase se encarga de crear las cuentas bancarias para los objetos de tipo
	 * persona y tienda. El atributo credenciales es un hashmap que como key recibe el id
	 * de la cuenta y como value el pin*/

package gestorAplicacion.gestionPersonas;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;

public class CuentaBancaria implements Serializable{

	//Atributo para el serializador
	private static final long serialVersionUID = 1L;
	private static ArrayList<CuentaBancaria> cuentasBancarias = new ArrayList<>();
	
	//Atributos
	private long nroCuenta;
	private float saldo;
	private int pin;
	static private HashMap<Long,Integer> credenciales = new HashMap<>();

	//Constructor

	public CuentaBancaria(long nroCuenta, float saldo, int pin) {
		this(nroCuenta,saldo);
		this.pin = pin;
		credenciales.put(nroCuenta, pin);
		cuentasBancarias.add(this);
	}
	//Sobrecarga de constructores que se utiliza para la clase vendedores
	public CuentaBancaria(long nroCuenta, float saldo){
		this.nroCuenta= nroCuenta;
		this.saldo=saldo;
		cuentasBancarias.add(this);
	}
	
	// Métodos
	
		/* Metodo para retirar dinero de la cuenta propia. Ingresamos el monto a retirar y el pin de la cuenta. 
		 * Se utiliza en el método transferir() */
		void retirar(float monto, int pin) {
			if (monto <= saldo & validarCredenciales(pin)==true) {
				setSaldo(getSaldo() - monto);
			}
		}
		
		/*Metodo para depositar dinero en la cuenta propia. Ingresamos el monto a depositar.
		 * Se utiliza en el método transferir() de la clase actual y en realizarCompra() de la clase Factura*/
		public void depositar(float monto) {
			setSaldo(monto + getSaldo());
		}

		/*Metodo para tranferir dinero a otra cuenta. Requiere ingresar la cuenta a la que se depositará 
		 * el dinero, el monto a depositar, y el pin de la cuenta*/
		public void tranferir(CuentaBancaria cuenta, float monto, int pin) {
			if (monto <= this.getSaldo() & validarCredenciales(pin)== true) {
				this.retirar(monto, pin);
				cuenta.depositar(monto);
			}
		}
		
		//Metodo para la seguridad al acceso del dinero de la cuenta
		public boolean validarCredenciales(int pin){
	        if (getPin() == pin){
	            return true;
	        } else {
	        	return false;
	        }
		}

	//Getter y setter utilizados para la serializacion y deserializacion de objetos
	public static ArrayList<CuentaBancaria> getCuentasBancarias() {
		return cuentasBancarias;
	}
	public static void setCuentasBancarias(ArrayList<CuentaBancaria> cuentasBancarias) {
		CuentaBancaria.cuentasBancarias=cuentasBancarias;
	}

	//Getters y setters
	public long getNroCuenta() {
		return nroCuenta;
	}

	public void setNroCuenta(int nroCuenta) {
		this.nroCuenta = nroCuenta;
	}

	public float getSaldo() {
		return saldo;
	}

	public void setSaldo(float saldo) {
		this.saldo = saldo;
	}
	
	public int getPin() {
		return pin;
	}
	
	public void setPin(int pin) {
		this.pin = pin;
	}

	public static HashMap<Long, Integer> getCredenciales() {
		return credenciales;
	}

	public static void setCredenciales(HashMap<Long, Integer> credenciales) {
		CuentaBancaria.credenciales = credenciales;
	}
}

File: Practica2-MascotasUN/src/gestorAplicacion/gestionPersonas/CuentaBancaria.py
class CuentaBancaria:
    
    _cuentasBancarias = []
    _credenciales = {}
    
    def __init__(self, nroCuenta=0, saldo=0, pin=0):
        self._nroCuenta = nroCuenta
        self._saldo = saldo
        self._pin = pin
        self._credenciales[nroCuenta] = pin
        self._cuentasBancarias.append(self)
        
    # Métodos
    
    # Metodo para retirar dinero de la cuenta propia. Ingresamos el monto a retirar y el pin de la cuenta. 
	# Se utiliza en el método transferir()
    def __str__(self) -> str:
        return  "~ NroCuenta: " + str(self._nroCuenta) + "\n" \
               + "~ Saldo: " + str(self._saldo) + "\n"
    def retirar(self, monto, pin):
        if monto <= self._saldo and self._validarCredenciales(pin) == True:
            self.setSaldo(self.getSaldo() - monto)
            
    # Metodo para depositar dinero en la cuenta propia. Ingresamos el monto a depositar.
	# Se utiliza en el método transferir() de la clase actual y en realizarCompra() de la clase Factura

    def depositar(self, monto):
        self.setSaldo(monto + self.getSaldo())
        
    # Metodo para tranferir dinero a otra cuenta. Requiere ingresar la cuenta a la que se depositará el dinero, 
    # el monto a depositar, y el pin de la cuenta
    
    def transferir(self, cuenta, monto, pin):
        if monto <= self.getSaldo() and self._validarCredenciales(pin) == True:
            self.retirar(monto, pin)
            if isinstance(cuenta, CuentaBancaria) == True:
                cuenta.depositar(monto)

    @staticmethod
    def crearObjeto(nroCuenta, saldo, pin):
        try:
            nroCuenta=int(nroCuenta)
            saldo = int(saldo)
            pin=int(pin)
            CuentaBancaria(nroCuenta, saldo, pin)
            return True
        except ValueError:
            return False

    # Getter y setter utilizados para la serializacion y deserializacion de objetos
    
    @classmethod
    def getCuentasBancarias(cls):
        return cls._cuentasBancarias
    
    @classmethod
    def setCuentasBancarias(cls, cuentasBancarias):
        cls._cuentasBancarias = cuentasBancarias
        
    # Getters y setters
        
    def getNroCuenta(self):
        return self._nroCuenta
    
    def setNroCuenta(self, nroCuenta):
        self._nroCuenta = nroCuenta
        
    def getSaldo(self):
        return self._saldo
    
    def setSaldo(self, saldo):
        self._saldo = saldo
        
    def getPin(self):
        return self._pin
    
    def setPin(self, pin):
        self._pin = pin
        
    @classmethod
    def getCredenciales(cls):
        return cls._credenciales
    
    @classmethod
    def setCredenciales(cls, credenciales):
        cls._credenciales = credenciales